baseclass/src/main/java/com/google/apigee/calloutbaseclass/CalloutBase.java

```diff
@@ -0,0 +1,191 @@
+  private static final String commonErrorPatternString = "^(.+?)[:;] (.+)$";
+  private static final Pattern commonErrorPattern = Pattern.compile(commonErrorPatternString);
+  // Matches any string between two curly braces, intended for flow variable references
+  // (e.g. matches "{request.test}", but also matches "{asdf}")
+  private static final String variableReferencePatternString =
+      "(.*?)\\{([^\\{\\} :][^\\{\\} ]*?)\\}(.*?)";
+  private static final Pattern variableReferencePattern =
+      Pattern.compile(variableReferencePatternString);
+
+  private final Map<String, String> properties;
+
+  /**
+   * Constructor taking in String key and value properties map
+   *
+   * @param properties String key and value map
+   */
+  public CalloutBase(Map<String, String> properties) {
+    this.properties = properties;
+  }
+
+  /**
+   * Retrieves an optional String property from properties map or empty if property does not exist
+   * in properties.
+   *
+   * @param propertyName Name of property to retrieve value
+   * @param messageContext Message Context
+   * @return Optional of property value in properties map or empty
+   */
+  public Optional<String> getOptionalProperty(String propertyName, MessageContext messageContext) {
+    if (!this.properties.containsKey(propertyName)) {
+      return Optional.empty();
+    }
+    String value =
+        resolveVariableReferences(this.properties.get(propertyName).trim(), messageContext);
+    if ("".equals(value)) {
+      return Optional.empty();
+    }
+    return Optional.of(value);
+  }
+
+  /**
+   * Retrieves a required property from properties map.
+   *
+   * @param propertyName Name of property to retrieve value
+   * @param messageContext Message Context
+   * @return Value of property in properties map
+   * @throws IllegalArgumentException if the propertyName does not exist in properties map or
+   *     resolves to an empty string.
+   */
+  public String getRequiredProperty(String propertyName, MessageContext messageContext) {
+    return getOptionalProperty(propertyName, messageContext)
+        .orElseThrow(IllegalArgumentException::new);
+  }
+
+  /**
+   * Resolves references to flow variables (i.e. variable references between two curly braces
+   * {request.example}) or the spec if not a variable reference.
+   *
+   * @param spec The potential flow variable reference
+   * @param messageContext Message Context
+   * @return Resolved flow variable reference or spec
+   */
+  public String resolveVariableReferences(String spec, MessageContext messageContext) {
+    Matcher matcher = variableReferencePattern.matcher(spec);
+    StringBuffer sb = new StringBuffer();
+    while (matcher.find()) {
+      matcher.appendReplacement(sb, "");
+      sb.append(matcher.group(1));
+      String ref = matcher.group(2);
+      String[] parts = ref.split(":", 2);
+      Object v = messageContext.getVariable(parts[0]);
+      if (v != null) {
```

Comment:
are both the if-block and the else-block tested?  I have a feeling that only one of them is.

Reply:
It'd help if you comment with an example what each block tries to catch.

Reply:
I'd provide an example where group 1, 2, 3 and part 1, 2 are all present.

Reply:
So the second part (split with ":") is supposed to be for a default value if the flow variable doesn't exist in the messageContext. Added a test for this.

Reply:
can you use a commentary to call this out explicit in this code?  That'd be very helpful for the readers in the future.